src/test/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/token/PersistentServiceProviderTokenStoreTest.java

```diff
@@ -0,0 +1,418 @@
+import static java.lang.System.currentTimeMillis;
+import static java.time.Duration.*;
+import static java.util.Optional.ofNullable;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+import static org.hamcrest.collection.IsMapContaining.hasEntry;
+import static org.hamcrest.collection.IsMapWithSize.aMapWithSize;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PersistentServiceProviderTokenStoreTest {
+
+    private static final ServiceProviderToken REQUEST_TOKEN_1 = newRequestToken("req-token1")
+            .callback(URI.create("http://some-callback-url/endpoint1"))
+            .consumer(RSA_CONSUMER)
+            .creationTime(currentTimeMillis())
+            .timeToLive(ofDays(1L).toMillis())
+            .tokenSecret("the-secret-token1")
+            .authorizedBy("test-user1")
+            .verifier("abc123")
+            .build();
+
+    private static final ServiceProviderToken REQUEST_TOKEN_2 = newRequestToken("req-token2")
+            .callback(URI.create("http://some-callback-url/endpoint2"))
+            .consumer(RSA_CONSUMER_WITH_2LO)
+            .creationTime(currentTimeMillis())
+            .timeToLive(ofDays(1L).toMillis())
+            .tokenSecret("the-secret-token2")
+            .authorizedBy("test-user2")
+            .verifier("def456")
+            .build();
+
+    private static final ServiceProviderToken ACCESS_TOKEN_1 = newAccessToken("access-token1")
+            .callback(URI.create("http://some-callback-url/endpoint1"))
+            .consumer(RSA_CONSUMER)
+            .creationTime(currentTimeMillis())
+            .timeToLive(ofDays(1L).toMillis())
+            .tokenSecret("the-secret-token3")
+            .authorizedBy("test-user2")
+            .session(newSession("session1")
+                    .creationTime(currentTimeMillis())
+                    .timeToLive(ofHours(1L).toMillis())
+                    .lastRenewalTime(currentTimeMillis())
+                    .build())
+            .verifier("xyz789")
+            .build();
+
+    private static final ServiceProviderToken ACCESS_TOKEN_2 = newAccessToken("access-token2")
+            .callback(URI.create("http://some-callback-url/endpoint3"))
+            .consumer(RSA_CONSUMER_WITH_2LO)
+            .creationTime(currentTimeMillis())
+            .timeToLive(ofDays(1L).toMillis())
+            .tokenSecret("the-secret-token4")
+            .authorizedBy("test-user3")
+            .session(newSession("session2")
+                    .creationTime(currentTimeMillis())
+                    .timeToLive(ofHours(1L).toMillis())
+                    .lastRenewalTime(currentTimeMillis())
+                    .build())
+            .verifier("mno567")
+            .build();
+
+    @Rule
+    public final JenkinsRule jenkins = new JenkinsRule();
```

Comment:
This is surely an integration test. This starts the entire Jenkins. We should move it to the correct place.

Reply:
It's... complicated. The `Secret` instances won't work without an instance and my github searches for sample code showed this trick for similar tests written for other plugins.
Happy to change if anyone has a different suggestion.

Reply:
I've observed this in other plugins as well, the line between unit and integration test is a bit blurry in Jenkins, given how much relies on a Jenkins instance running. We have a `JenkinsProvider` for mocking Jenkins, but that won't help here. I think I'd still argue they should be labelled as integration tests though (and just accept unit testing is probably impossible :( )

Reply:
> I've observed this in other plugins as well, the line between unit and integration test is a bit blurry
I don't think it is blurry at all. If you are starting Jenkins as part of the test, its not a unit test anymore. Quantifiable definition of a unit test class is that it should run in < 1 second. The ideal is 500ms. Anything more is integration tests. If you are happy with the coverage, we can keep this `it.com.atlassian.bitbucket.jenkins.internal` and not have a unit test case.

Reply:
I agree there are shades of grey with these things, and it seems to me we're discussing semantics here.
I don't disagree that this is less than ideal (unfortunately, most things with Jenkins seem to be), but IMO this is _not_ an integration test or even func-test _just_ because it starts up an instance of Jenkins, or it takes more than 500ms (although those would be very good objectives to pursue in a unit test). This test doesn't verify real implementations of multiple parts and components of the system working together, i.e. this doesn't test integration.
This tests a _unit_ in the system (the `TokenStore`) and uses mocks and/or stubs for the dependent components (like the `ConsumerStore`)...
I think a good way to try to bypass the need for starting up Jenkins is to redesign the class so that a mock of the `Secret` dependency can be injected instead. I'll see if I can do that.
But I would like to repeat: IMO, this is _not_ an integration test just because we have to spin up Jenkins, that's just an unfortunate side-effect of the way `Secret` is implemented (which tbf, may have been inevitable for them).

Reply:
P.S. I just ran it again in Intellinj and the test takes about 10 seconds which is definitely unacceptable for a unit test. :+1: 
I'll work on bypassing the Jenkins rule and mocking `Secret` somehow.

Reply:
Your test will fail if BitbucketSCM has some runtime problem in its logic. When this failure happens, you will have to look into the detail to know what went wrong. BitbucketSCM is as far a functionality is from this as any can be. If you don't have specific asserts regarding this functionality doesn't mean you don't depend on it. Unit test and ideally your submodule should work irrespective of failure in BitbucketSCM. This is not a semantic thing. I was only asking for it to be moved to the correct integration package. 

We expect the feedback from the build + unit test to be fast. Imagine having 20 tests like this.

Reply:
I thought a bit more about this and I agree we're definitely entering the realm of integration tests here, because we're _actually_ marshalling and unmarshalling the tokens to/from XML and actually writing to disk (test/temp file) and reading from disk, instead of abstracting and mocking/stubbing the persistence part of the functionality. :+1: 
I'll be happy to move this to the IT package and probably make it less fine-grained (e.g. we just need to make sure the marshalling/unmarshalling works properly with the happy path when we call `load()` and `save()`), and then once I extract and abstract the XML/persistence handling to a separate class, we can have a fine-grained unit test that tests against the mock of that interface/class.